src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java

```diff
@@ -515,26 +520,63 @@ public void afterAnalysis(
-      for (ConfiguredTarget configuredTarget : configuredTargets) {
+      Set<ActionInput> filesToDownload = new HashSet<>();
+      for (ConfiguredTarget configuredTarget : analysisResult.getTargetsToBuild()) {
         if (isTestCommand && isTestRule(configuredTarget)) {
-          // When running a test download the test.log and test.xml.
+          // When running a test download the test.log and test.xml. These are never symlinks.
           filesToDownload.addAll(getTestOutputs(configuredTarget));
         } else {
-          filesToDownload.addAll(getArtifactsToBuild(configuredTarget, artifactContext).toList());
-          filesToDownload.addAll(getRunfiles(configuredTarget));
+          fetchSymlinkDependenciesRecursively(
+              analysisResult.getActionGraph(),
+              filesToDownload,
+              getArtifactsToBuild(configuredTarget, artifactContext).toList());
+          fetchSymlinkDependenciesRecursively(
+              analysisResult.getActionGraph(),
+              filesToDownload,
+              getRunfiles(configuredTarget));
         }
       }
-      actionContextProvider.setFilesToDownload(filesToDownload.build());
+      actionContextProvider.setFilesToDownload(ImmutableSet.copyOf(filesToDownload));
+    }
+  }
+
+  // This is a short-term fix for top-level outputs that are symlinks. Unfortunately, we cannot
+  // reliably tell after analysis whether actions will create symlinks (the RE protocol allows any
+  // action to generate and return symlinks), but at least we can handle basic C++ rules with this
+  // change.
+  // TODO(ulfjack): I think we should separate downloading files from action execution. That would
+  // also resolve issues around action invalidation - we currently invalidate actions to trigger
+  // downloads of top-level outputs when the top-level targets change.
+  private static void fetchSymlinkDependenciesRecursively(
+      ActionGraph actionGraph, Set<ActionInput> builder, List<Artifact> inputs) {
+    for (Artifact input : inputs) {
+      // Only fetch recursively if we don't have the file to avoid visiting nodes multiple times.
+      if (builder.add(input)) {
+        fetchSymlinkDependenciesRecursively(actionGraph, builder, input);
+      }
+    }
+  }
+
+  private static void fetchSymlinkDependenciesRecursively(
+      ActionGraph actionGraph, Set<ActionInput> builder, Artifact artifact) {
+    if (!(actionGraph.getGeneratingAction(artifact) instanceof ActionExecutionMetadata)) {
+      // The top-level artifact could be a tree artifact, in which case the generating action may
+      // be an ActionTemplate, which does not implement ActionExecutionMetadata. We don't handle
+      // this case right now, so exit.
+      return;
+    }
+    ActionExecutionMetadata action = (ActionExecutionMetadata) actionGraph.getGeneratingAction(artifact);
+    if (action.mayInsensitivelyPropagateInputs()) {
+      List<Artifact> inputs = action.getInputs().toList();
```

Comment:
Would you mind adding a warning-level logline here if the list is >5 elements or so? I don't think we expect such actions to insensitively propagate inputs, and if there's a bug, we could slow down significantly due to nested set expansion.

Reply:
Done.